### PR TITLE
Will close #269 at some point

### LINKED ISSUE #269: Make open() handle binary files
I'm thinking `open(filename, "b")`. Should return a wrapper type with a `.length` and `.hex()` at the very least, and functions in `k6/crypto`, etc. should be amended to handle these values.

### PR DESCRIPTION
**Pull request by ragnarlonn**
A PR for you to tear to pieces, o great maintainer

### CODE REVIEW

**Comment by ragnarlonn** [comment]
Note: the `k6/encoding` module should also be changed to recognize the TextOrBinary object as input parameter

**Comment by liclac** [comment]
TIL you can just change the parameters' type to `[]byte` and it'll Just Work™. There's a weird edge case where passing `[]byte("hi")` to a `string` parameter turns it into a string that says "104,105", I'm gonna open a goja issue about this.

**Comment by ragnarlonn** [comment]
Which parameter are you talking about?  And apologies if it is obvious - my brain is a bit woolly today (evil cold virus).

**Comment by liclac** [comment]
Any `string` parameter that should be able to accept bytes. I'm discussing a better solution with the goja maintainer, but this will work as a stopgap.

**Comment by ragnarlonn** [comment]
@liclac So, I'm still a bit confused about exactly what you'd like me to change. Do you mean that the parameters to the crypto functions should be all []byte instead of goja.Value?  That would mean you can't pass the binary data object you get back from open() directly to the crypto functions (or any other functions that want to consume binary data), but would have to do e.g. `let dat = open("file.dat", "b");  hasher.update(dat.bytes());` instead of `let dat = open("file.dat", "b"); hasher.update(dat);`  But maybe that is better?

**Comment by ragnarlonn** [comment]
ICMP echo request -> @liclac

**Comment by liclac** [comment]
Oh, my response didn't send.

As https://github.com/dop251/goja/issues/51 points out, we should either implement an `ArrayBuffer`-like (which we can PR to goja), or do the quick fix which is to pass around `[]byte` objects, which will Just Work™.

**Comment by ragnarlonn** [comment]
@liclac OK, I had some idea that it would be bad if we didn't know whether data was binary or not - that we might want other functions (like perhaps the crypto.* ones) to know this. If we just use a []byte parameter to pass it around we don't know if it is a legible string, or something that contains null bytes. But I can't really think of an example where it would give us trouble. I'll try just using byte slices everywhere and see what happens!

**Comment by liclac** [comment]
The easiest solution would probably be to let the `console.*` functions explicitly check if `v.ExportType()` is a `[]byte` and cast to string in those cases, otherwise do `v.String()` like normal.

File: js/console.go [hunk 1]
```diff
@@ -47,9 +48,16 @@ func (c Console) log(ctx *context.Context, level log.Level, msgobj goja.Value, a
 
 	fields := make(log.Fields)
 	for i, arg := range args {
-		fields[strconv.Itoa(i)] = arg.String()
+		if arg.ExportType() == reflect.TypeOf((*[]byte)(nil)).Elem() {
```
Comment (liclac) [inline js/console.go]: You can declare a global `var typeBytes = reflect.TypeOf(...)` to not have to recalculate this every time.

**Comment by robingustafsson** [comment]
@ragnarlonn When will you be able to fix the merge conflict and move the type calculation that @liclac commented on in console.go? I want to get this merged so we can move https://github.com/loadimpact/k6/pull/420 forward.

**Comment by robingustafsson** [comment]
Closing this in favor of https://github.com/loadimpact/k6/pull/524.

**Pull request closed without merging**